arch/x86/kvm/pkvm/mmu.c

```
@@ -612,3 +612,124 @@ int pkvm_hyp_donate_host_mmio_locked(unsigned long phys, unsigned long size)
+			}
+
+			/*
+			 * Allow sharing an already shared page as long as it is
+			 * shared with the hypervisor, not with a guest.
+			 */
+			if (page->host_share_hyp_count)
+				continue;
+
+			fallthrough;
+		default:
+			ret = -EPERM;
+			goto unlock;
+		}
+	}
+
+	for (phys = start; phys < end; phys += PAGE_SIZE) {
+		page = pkvm_phys_to_page(phys);
+
+		page->host_state = PKVM_PAGE_SHARED_OWNED;
+		page->host_share_hyp_count++;
+	}
+
+	ret = 0;
+
+unlock:
+	pkvm_host_mmu_unlock();
+
+	return ret;
+}
+
+/**
+ * pkvm_host_unshare_hyp() - Host un-shares pages with hypervisor.
+ * @phys:	Physical address of the memory region to unshare.
+ * @size:	Size of the memory region to unshare.
+ *
+ * Decrements the host_share_hyp_count counter for the memory pages in range
+ * [@phys, @phys + @size) which are shared-owned owned by the host and the pKVM
+ * hypervisor, and if a page's host_share_hyp_count becomes zero, then changes
+ * this page's state to make it exclusively owned by the host. @phys and @size
+ * are not required to be PAGE_SIZE aligned (@size is required to be non-zeroed
+ * value), to support unsharing arbitrary unaligned memory regions.
+ *
+ * Returns: 0 on success, or a negative error code on failure.
+ */
+int pkvm_host_unshare_hyp(unsigned long phys, unsigned long size)
+{
+	unsigned long end = PAGE_ALIGN(phys + size);
+	struct pkvm_page *page;
+	int ret;
+
+	if (size == 0)
+		return -EINVAL;
+
+	pkvm_host_mmu_lock();
+
+	ret = check_host_mem_pgstate(phys, size, PKVM_PAGE_SHARED_OWNED);
+	if (ret)
+		goto unlock;
+
+	for (phys = PAGE_ALIGN_DOWN(phys); phys < end; phys += PAGE_SIZE) {
+		page = pkvm_phys_to_page(phys);
+
+		/*
+		 * Even if host_share_hyp_count is 0 because the page is
+		 * shared with a guest, not with the hypervisor, it still
+		 * means a pKVM bug, since pkvm_host_unshare_hyp() is only
+		 * used by pKVM itself, for pages that are known to be
+		 * shared with the hypervisor.
+		 */
+		BUG_ON(!page->host_share_hyp_count);
```

Just a note: I suggested this BUG_ON myself, but actually we could go ahead and implement a logic similar to the logic in pkvm_host_share_hyp(), i.e. first check page states and host_share_hyp_count counters for all pages, and then (if check succeeded) update page states and counters in a separate loop, to avoid this BUG_ON and thus make pkvm_host_unshare_hyp() more robust (and more consistent: in case of check_host_mem_pgstate() failure above we return an error instead of BUG_ON, so for consistency we could return an error in this case as well?).

But I'm probably fine either way.

---

My understanding is that, for the `pkvm_host_share_hyp()`, we prefer the logic to first check page states and `host_share_hyp_count` for all pages in a for loop is because `pkvm_host_share_hyp()` is taking the input address from the host which is untrusted. So it is something can happen even there is no code bug at all in the pKVM.

For `pkvm_host_unshare_hyp()`, I cannot imagine if there is any case for this function to take the input address from the host. If so, it only takes the input address from the pKVM hypervisor, which is trusted. Then it is something shouldn't happen if we have no bug. Adding a similar logic to check `host_share_hyp_count` for unshare seems almost not quite useful as most likely we don't have such bug?

To me, this BUG_ON takes the same effect with the BUG_ON in the first for loop of the `pkvm_host_share_hyp()`, which is purely to make sure we don't have any code bug in the pKVM hypervisor.

From this point of view, I prefer we keep to use the BUG_ON to keep it simpler.

---

Yeah, from this point of view it makes sense.

And this actually makes me think: would it be even better to add BUG_ON for the above check_host_mem_pgstate() check as well, and for `size == 0`, and make pkvm_host_unshare_hyp() void?

---

Yeah, also makes sense to me.